jazz_cognito-authorizer/index.js

```diff
@@ -165,330 +125,370 @@ function ValidateToken(pems, event, context, cb) {
+         * @param {Object} The conditions object in the format specified by the AWS docs.
          * @return {void}
          */
-        allowMethodWithConditions: function(verb, resource, conditions) {
-            addMethod.call(this, "allow", verb, resource, conditions);
-        },
+        var addMethod = function (effect, verb, resource, conditions) {
+            if (verb != "*" && !AuthPolicy.HttpVerb.hasOwnProperty(verb)) {
+                throw new Error("Invalid HTTP verb " + verb + ". Allowed verbs in AuthPolicy.HttpVerb");
+            }
+
+            if (!this.pathRegex.test(resource)) {
+                throw new Error("Invalid resource path: " + resource + ". Path should match " + this.pathRegex);
+            }
+
+            var cleanedResource = resource;
+            if (resource.substring(0, 1) == "/") {
+                cleanedResource = resource.substring(1, resource.length);
+            }
+            var resourceArn = "arn:aws:execute-api:" +
+                this.region + ":" +
+                this.awsAccountId + ":" +
+                this.restApiId + "/" +
+                this.stage + "/" +
+                verb + "/" +
+                cleanedResource;
+
+            if (effect.toLowerCase() == "allow") {
+                this.allowMethods.push({
+                    resourceArn: resourceArn,
+                    conditions: conditions
+                });
+            } else if (effect.toLowerCase() == "deny") {
+                this.denyMethods.push({
+                    resourceArn: resourceArn,
+                    conditions: conditions
+                });
+            }
+        };
 
         /**
-         * Adds an API Gateway method (Http verb + Resource path) to the list of denied
-         * methods and includes a condition for the policy statement. More on AWS policy
-         * conditions here: http://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements.html#Condition
+         * Returns an empty statement object prepopulated with the correct action and the
+         * desired effect.
          *
-         * @method denyMethodWithConditions
-         * @param {String} The HTTP verb for the method, this should ideally come from the
-         *                 AuthPolicy.HttpVerb object to avoid spelling mistakes
-         * @param {string} The resource path. For example "/pets"
-         * @param {Object} The conditions object in the format specified by the AWS docs
-         * @return {void}
+         * @method getEmptyStatement
+         * @param {String} The effect of the statement, this can be "Allow" or "Deny"
+         * @return {Object} An empty statement object with the Action, Effect, and Resource
+         *                  properties prepopulated.
          */
-        denyMethodWithConditions : function(verb, resource, conditions) {
-            addMethod.call(this, "deny", verb, resource, conditions);
-        },
+        var getEmptyStatement = function (effect) {
```

Comment:
Looks like some good amount of refactoring. Have you tested all of these changes?